xsimlab/progress.py

```diff
@@ -0,0 +1,88 @@
+from xsimlab.hook import RuntimeHook, runtime_hook
+from xsimlab.utils import format_time
+
+
+class ProgressBar(RuntimeHook):
+    """
+    Progress bar implementation using the tqdm package.
+
+    Parameters
+    ----------
+    frontend : {"auto", "console", "gui", "notebook"}, optional
+        Allows control over Python environment.
+    **kwargs : dict, optional
+        Arbitrary keyword arguments for progress bar customization.
+
+    Examples
+    --------
+    :class:`ProgressBar` takes full advantage of :class:`RuntimeHook`.
+
+    Call it as part of :func:`run`:
+    >>> out_ds = in_ds.xsimlab.run(model=model, hooks=[xs.ProgressBar()])
+
+    In a context manager using the `with` statement`:
+    >>> with xs.ProgressBar():
+    ...    out_ds = in_ds.xsimlab.run(model=model)
+
+    Globally with `register` method:
+    >>> pbar = xs.ProgressBar()
+    >>> pbar.register()
+    >>> out_ds = in_ds.xsimlab.run(model=model)
+    >>> pbar.unregister()
+
+    For additional customization, see: https://tqdm.github.io/docs/tqdm/
+    """
+
+    def __init__(self, frontend="auto", **kwargs):
+        if frontend == "auto":
+            from tqdm.auto import tqdm
+        elif frontend == "console":
+            from tqdm import tqdm
+        elif frontend == "gui":
+            from tqdm.gui import tqdm
+        elif frontend == "notebook":
+            from tqdm.notebook import tqdm
+        else:
+            raise ValueError(
+                f"Frontend argument {frontend!r} not supported. Please select one of the following: {', '.join(['auto', 'console', 'gui', 'notebook'])}"
+            )
+
+        self.custom_description = True
+        if not "desc" in kwargs.keys():
+            self.custom_description = False
```

Comment:
nit: testing the opposite is slightly clearer IMO.

```suggestion
        self.custom_description = False
        if "desc" in kwargs.keys():
            self.custom_description = True
```